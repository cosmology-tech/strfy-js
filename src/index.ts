export * from './js';
export * from './json';