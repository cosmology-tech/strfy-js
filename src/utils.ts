import { minimatch } from 'minimatch';

export function camelCaseTransform(key: string): string {
  return key.replace(/[-_\s]+(.)?/g, (_, c) => c ? c.toUpperCase() : '');
}

export function isSimpleKey(key: string): boolean {
  return /^[a-zA-Z_$][a-zA-Z0-9_$]*$/.test(key);
}

export function dirname(path: string): string {
  return path.replace(/\/[^\/]*$/, ''); // Removes last segment after the last '/'
}

export function escapeStringForSingleQuotes(str: string): string {
  // Escape backslashes first
  str = str.replace(/\\/g, '\\\\');
  // Escape control characters
  str = str.replace(/\n/g, '\\n');
  str = str.replace(/\r/g, '\\r');
  str = str.replace(/\t/g, '\\t');
  // Escape only single quotes
  str = str.replace(/'/g, "\\'");
  return str;
}

export function escapeStringForDoubleQuotes(str: string): string {
  // Escape backslashes first
  str = str.replace(/\\/g, '\\\\');
  // Escape control characters
  str = str.replace(/\n/g, '\\n');
  str = str.replace(/\r/g, '\\r');
  str = str.replace(/\t/g, '\\t');
  // Escape only double quotes
  str = str.replace(/"/g, '\\"');
  return str;
}


export function escapeStringForBacktickQuotes(str: string): string {
  // Escape backslashes first
  str = str.replace(/\\/g, '\\\\');
  // Escape control characters
  str = str.replace(/\n/g, '\\n');
  str = str.replace(/\r/g, '\\r');
  str = str.replace(/\t/g, '\\t');
  // Escape backticks
  str = str.replace(/`/g, '\\`');
  return str;
}


const globPattern = /\*+([^+@!?\*\[\(]*)/;

interface ShouldIncludeOptions {
  include: string[];
  exclude: string[];
}

export function expandPaths(paths: string[]): string[] {
  const pathSet = new Set<string>();

  paths.forEach(path => {
    let currentPath = path;
    // Ensure the path is added to the set
    pathSet.add(currentPath);

    // Generate and add all parent directories to the set
    while (currentPath !== '') {
      const lastSlashIndex = currentPath.lastIndexOf('/');
      if (lastSlashIndex === -1) break; // No more slashes, stop processing

      currentPath = currentPath.substring(0, lastSlashIndex) || '';
      if (currentPath !== '')
        pathSet.add(currentPath);
    }
  });

  return sortPaths(Array.from(pathSet));
}

function sortPaths(paths) {
  // First sort alphabetically
  paths.sort((a, b) => a.localeCompare(b));

  // Then sort by depth (number of '/' in the path)
  paths.sort((a, b) => {
    const depthA = (a.match(/\//g) || []).length;
    const depthB = (b.match(/\//g) || []).length;
    return depthB - depthA;  // Sort from deepest to shallowest
  });

  return paths;
}


function hasAncestor(currentPath, includePaths) {
  return !!findNearestAncestor(currentPath, includePaths);
}

function findNearestAncestor(currentPath, includePaths) {
  let pathToCheck = currentPath;
  while (pathToCheck !== '/' && pathToCheck !== '') {
    if (includePaths.some(includePath => minimatch(pathToCheck + '/', includePath))) {
      if (pathToCheck !== currentPath)
        return pathToCheck; // Return the matching parent directory
    }
    pathToCheck = dirname(pathToCheck); // Move to the next parent directory
  }
  return null; // Return null if no matching parent is found
}

export const shouldInclude = (type: string, options: ShouldIncludeOptions): boolean => {
  // Determine if 'include' and 'exclude' are effectively set
  const includesEffectivelySet = options.include && options.include.length > 0;
  const excludesEffectivelySet = options.exclude && options.exclude.length > 0;

  // Function to check if any patterns in the array match the type
  const matchesPattern = (patterns: string[], type: string): boolean =>
    patterns.some(pattern => globPattern.test(pattern) ? minimatch(type, pattern) : type === pattern);

  const isIncluded = includesEffectivelySet ? matchesPattern(expandPaths(options.include), type) : true;
  const isIncludedByAncestor = !!findNearestAncestor(type, options.include);
  const isExcluded = excludesEffectivelySet ? matchesPattern(options.exclude, type) : false;

  const toInclude = isIncluded;

  // Apply the logic based on whether includes or excludes are effectively set
  if (includesEffectivelySet && excludesEffectivelySet) {
    if (isIncludedByAncestor && !isExcluded) {
      return true;
    }
    return toInclude && !isExcluded;
  } else if (includesEffectivelySet) {
    if (isIncludedByAncestor) {
      return true;
    }
    return toInclude;
  } else if (excludesEffectivelySet) {
    return !isExcluded;
  }

  // Default behavior if neither is effectively set
  return true;
}