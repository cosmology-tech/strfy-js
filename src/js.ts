import { camelCaseTransform, dirname, escapeStringForBacktickQuotes, escapeStringForDoubleQuotes, escapeStringForSingleQuotes, isSimpleKey, shouldInclude } from "./utils";

export interface JSStringifySetterOptions<ObjType, RootType> {
  currentKey: string;
  currentPath: string;
  value: any;
  obj: ObjType;
  root: RootType;
  defaultValuesMap: { [path: string]: any };
};

export interface JSStringifyPropertyReplacerOptions<ObjType, RootType> {
  currentKey: string;
  currentPath: string;
  propertyRenameMap: { [pattern: string]: string };
  obj: ObjType;
  root: RootType
  value?: any;
}

export type JSStringifySetter = (options: JSStringifySetterOptions<any, any>) => any;
export type JSStringifyReplacer = (options: JSStringifyPropertyReplacerOptions<any, any>) => any;
export interface JSStringifyOptions {
  space?: number;
  propertyReplacer?: (options: JSStringifyPropertyReplacerOptions<any, any>) => string
  valueReplacer?: Record<string, (options: JSStringifyPropertyReplacerOptions<any, any>) => any>;
  quotes?: 'single' | 'double' | 'backtick';
  inlineArrayLimit?: number;
  camelCase?: boolean;
  camelCaseFn?: (str: string) => string;
  include?: string[];
  exclude?: string[];
  defaultValuesMap?: { [path: string]: any };
  propertyRenameMap?: { [path: string]: any };
  defaultValuesSetter?: { [path: string]: JSStringifySetter } | JSStringifySetter;
  json?: boolean;
}


export function chooseQuotes(str: string, preferred: 'single' | 'double' | 'backtick'): string {
  switch (preferred) {
    case 'single':
      return `'${escapeStringForSingleQuotes(str)}'`;
    case 'double':
      return `"${escapeStringForDoubleQuotes(str)}"`;
    case 'backtick':
      return `\`${escapeStringForBacktickQuotes(str)}\``;
    default:
      throw new Error("Invalid quote type specified.");
  }
}

function matchesPattern(path: string, pattern: string) {
  if (pattern === '*') return true;
  if (path === pattern) return true;
  const pathSegments = path.split('/');
  const patternSegments = pattern.split('/');
  if (pathSegments.length !== patternSegments.length) {
    return false;
  }
  for (let i = 0; i < pathSegments.length; i++) {
    if (patternSegments[i] !== '*' && patternSegments[i] !== pathSegments[i]) {
      return false;
    }
  }
  return true;
}

function defaultPropertyReplacer(
  currentKey: string,
  currentPath: string,
  propertyRenameMap: { [pattern: string]: string }
): string {
  let finalKey = currentKey;
  Object.keys(propertyRenameMap).forEach(pattern => {
    if (matchesPattern(currentPath, pattern)) {
      finalKey = propertyRenameMap[pattern];
    }
  });
  return finalKey;
}

export function jsStringify(obj: any, options?: JSStringifyOptions): string {
  const {
    space = 0,
    propertyReplacer = null,
    valueReplacer = {},
    quotes = 'single',
    inlineArrayLimit = undefined,
    camelCase = false,
    camelCaseFn = camelCaseTransform,
    propertyRenameMap = {},
    defaultValuesMap = {},
    defaultValuesSetter = {},
    include = [],
    exclude = [],
    json = false  // JSON compliance mode
  } = options || {};

  const isObject = (val: any): boolean => typeof val === 'object' && val !== null && !Array.isArray(val);

  let indentLevel: number = 0;

  const serialize = (root: any, obj: any, currentPath = ""): string => {
    if (valueReplacer[''] instanceof Function) {
      obj = valueReplacer['']({
        currentKey: '',
        currentPath,
        obj,
        propertyRenameMap,
        root,
        value: obj
      }); // Call valueReplacer for the root element initially
    }

    if (Array.isArray(obj)) {
      const useInline = inlineArrayLimit !== undefined && obj.length <= inlineArrayLimit;
      indentLevel++;
      const result = '[' + (space && !useInline ? '\n' : '') + obj.map((item, index) => {
        return ' '.repeat(useInline ? 0 : indentLevel * space) + serialize(root, item, `${currentPath}/${index}`);
      }).join(',' + (space && !useInline ? '\n' : ' ')) + (space && !useInline ? '\n' + ' '.repeat((indentLevel - 1) * space) : '') + ']';
      indentLevel--;
      return result;
    } else if (isObject(obj)) {
      indentLevel++;

      // Apply custom default values setter if available
      Object.keys(valueReplacer).forEach(pattern => {
        if (matchesPattern(currentPath, dirname(pattern))) {
          const replacer: JSStringifyReplacer = valueReplacer[pattern];
          const property = pattern.split('/').pop(); // get the last segment as property name
          if (property === '*') {
            // Apply setter to all properties if needed
            Object.keys(obj).forEach(prop => {
              obj[prop] = replacer({
                currentKey: prop,
                currentPath: `${currentPath}/${prop}`,
                obj,
                value: obj[prop],
                root,
                propertyRenameMap
              });
            });
          } else if (property) {
            obj[property] = replacer({
              currentKey: property,
              currentPath,
              obj,
              value: obj[property],
              root,
              propertyRenameMap
            });
          }

        }
      });

      // Apply custom default values setter if available
      Object.keys(defaultValuesSetter).forEach(pattern => {
        if (matchesPattern(currentPath, dirname(pattern))) {
          const setter: JSStringifySetter = defaultValuesSetter[pattern];
          const property = pattern.split('/').pop(); // get the last segment as property name
          if (property === '*') {
            // Apply setter to all properties if needed
            Object.keys(obj).forEach(prop => {
              if (obj[prop] === undefined) {
                obj[prop] = setter({
                  currentKey: prop,
                  currentPath: `${currentPath}/${prop}`,
                  obj,
                  value: obj[prop],
                  root,
                  defaultValuesMap
                });
              }
            });
          } else if (property && obj[property] === undefined) {
            obj[property] = setter({
              currentKey: property,
              currentPath,
              obj,
              value: obj[property],
              root,
              defaultValuesMap
            });
          }
        }
      });

      // Apply default values based on currentPath
      Object.keys(defaultValuesMap).forEach(pattern => {
        if (matchesPattern(currentPath, pattern)) {
          const property = pattern.split('/').pop(); // get the last segment as property name
          if (property === '*') {
            // Apply to all properties if needed
            Object.keys(obj).forEach(prop => {
              if (obj[prop] === undefined) {
                obj[prop] = defaultValuesMap[pattern];
              }
            });
          } else if (property && obj[property] === undefined) {
            obj[property] = defaultValuesMap[pattern];
          }
        }
      });


      // INCLUDES EXCLUDES

      obj = Object.keys(obj).reduce((m, key) => {
        const fullPath = `${currentPath}/${key}`;
        if (shouldInclude(fullPath, {
          exclude,
          include
        })) {
          m[key] = obj[key];
        }
        return m;
      }, {});


      // CHANGE PROPERTY NAMES

      const props = Object.keys(obj).map(key => {
        const fullPath = `${currentPath}/${key}`;
        const value = valueReplacer instanceof Function ? valueReplacer({
          currentKey: key,
          currentPath: fullPath,
          obj: obj[key],
          propertyRenameMap,
          root
        }) : obj[key];

        let replacedKey: string;
        switch (true) {
          case propertyReplacer instanceof Function:
            replacedKey = propertyReplacer({ root, obj, currentKey: key, currentPath: fullPath, propertyRenameMap });
            break;
          default:
            replacedKey = defaultPropertyReplacer(key, fullPath, propertyRenameMap);
        }

        const finalKey = camelCase ? camelCaseFn(replacedKey) : replacedKey;
        const keyPart = json ? `"${finalKey}"` : isSimpleKey(finalKey) ? finalKey : `"${finalKey}"`;
        const valuePart = serialize(root, value, fullPath);
        return ' '.repeat(indentLevel * space) + `${keyPart}: ${valuePart}`;
      });
      const result = '{' + (space ? '\n' : '') + props.join(',' + (space ? '\n' : ' ')) + (space ? '\n' + ' '.repeat((indentLevel - 1) * space) : '') + '}';
      indentLevel--;
      return result;
    } else if (typeof obj === 'string') {
      return json ? `"${escapeStringForDoubleQuotes(obj)}"` : chooseQuotes(obj, quotes);
    } else {
      return String(obj);
    }
  };

  return serialize(obj, obj);
}
