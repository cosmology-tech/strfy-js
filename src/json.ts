import { camelCaseTransform, isSimpleKey } from "./utils";

export interface JSONStringifyOptions {
  space?: number; // defaults to 0
  replacer?: (this: any, key: string, value: any) => any; // defaults to null
  quotes?: 'single' | 'double'; // defaults to double
  inlineArrayLimit?: number; // default is undefined, allowing inline arrays below this length
  camelCase?: boolean; // defaults to false
  camelCaseFn?: (str: string) => string; // optional function to convert keys to camelCase
}

function escapeString(str: string): string {
  str = str.replace(/\\/g, '\\\\');  // Escape backslashes first
  str = str.replace(/\n/g, '\\n');  // Escape newlines
  str = str.replace(/\r/g, '\\r');  // Escape carriage returns
  str = str.replace(/\t/g, '\\t');  // Escape tabs
  str = str.replace(/"/g, '\\"');  // Escape double quotes
  return str;
}

function chooseQuotes(str: string): string {
  return `"${escapeString(str)}"`;
}

export function jsonStringify(obj: any, options?: JSONStringifyOptions): string {
  const { space = 0, replacer = null, quotes = 'double', inlineArrayLimit = undefined, camelCase = false, camelCaseFn = camelCaseTransform } = options || {};
  const isObject = (val: any): boolean => typeof val === 'object' && val !== null && !Array.isArray(val);
  let indentLevel: number = 0;

  const serialize = (obj: any): string => {
    if (replacer instanceof Function) {
      obj = replacer.call(null, '', obj); // Call replacer for the root element initially
    }

    if (Array.isArray(obj)) {
      const useInline = inlineArrayLimit !== undefined && obj.length <= inlineArrayLimit;
      indentLevel++;
      const result = '[' + (space && !useInline ? '\n' : '') + obj.map(item => {
        return ' '.repeat(useInline ? 0 : indentLevel * space) + serialize(item);
      }).join(',' + (space && !useInline ? '\n' : ' ')) + (space && !useInline ? '\n' + ' '.repeat((indentLevel - 1) * space) : '') + ']';
      indentLevel--;
      return result;
    } else if (isObject(obj)) {
      indentLevel++;
      const props = Object.keys(obj).map(key => {
        const value = replacer instanceof Function ? replacer.call(obj, key, obj[key]) : obj[key];
        const transformedKey = camelCase ? camelCaseFn(key) : key;
        const keyPart = isSimpleKey(transformedKey) ? transformedKey : chooseQuotes(transformedKey);
        const valuePart = serialize(value);
        return ' '.repeat(indentLevel * space) + `${keyPart}: ${valuePart}`;
      });
      const result = '{' + (space ? '\n' : '') + props.join(',' + (space ? '\n' : ' ')) + (space ? '\n' + ' '.repeat((indentLevel - 1) * space) : '') + '}';
      indentLevel--;
      return result;
    } else if (typeof obj === 'string') {
      return chooseQuotes(obj);
    } else {
      return String(obj);
    }
  };

  return serialize(obj);
}
