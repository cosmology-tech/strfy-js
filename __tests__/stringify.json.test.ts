import { jsonStringify, JSStringifyPropertyReplacerOptions } from '../src';

it('serializes simple objects without quotes on keys where possible', () => {
  const obj = {
    id: 1,
    name: 'Alice',
    isActive: true
  };
  const output = jsonStringify(obj);
  expect(output).toMatchSnapshot();
});

it('handles complex nested objects with arrays', () => {
  const obj = {
    user: {
      id: 2,
      details: {
        name: 'Bob',
        hobbies: ['skiing', 'cycling']
      }
    }
  };
  const output = jsonStringify(obj, { space: 2 });
  expect(output).toMatchSnapshot();
});

it('uses single quotes for strings by default', () => {
  const obj = {
    greeting: "Hello, world!"
  };
  const output = jsonStringify(obj);
  expect(output).toMatchSnapshot();
});

it('switches to backticks when single quotes are in the string', () => {
  const obj = {
    message: "It's a wonderful day!"
  };
  const output = jsonStringify(obj);
  expect(output).toMatchSnapshot();
});

it('uses double quotes when backticks and single quotes are present', () => {
  const obj = {
    quote: "`This` is 'awesome'!"
  };
  const output = jsonStringify(obj, { quotes: 'double' });
  expect(output).toMatchSnapshot();
});

it('properly escapes strings when necessary', () => {
  const obj = {
    complexString: "She said, \"That's `incredible`!\""
  };
  const output = jsonStringify(obj, { quotes: 'double' });
  expect(output).toMatchSnapshot();
});

interface Person {
  firstName: string;
  lastName: string;
  age: number
}

it('applies replacer function if provided', () => {
  const obj: Person[] = [
    {
      firstName: 'Alice',
      lastName: 'Johnson',
      age: 30
    },
    {
      firstName: 'Dana',
      lastName: 'Johnson',
      age: 30
    }
  ]
  const output = jsonStringify(obj, {
    valueReplacer: {
      '*': (opts: JSStringifyPropertyReplacerOptions<Person, Person[]>) => {
        if (opts.obj.lastName === 'Johnson') {
          if (opts.currentKey === 'age') {
            return 20;
          }
        }
        return opts.value;
      },
      '/*/age': (opts: JSStringifyPropertyReplacerOptions<Person, Person[]>) => {
        if (opts.obj.firstName === 'Dana') return 46;
        return opts.value;
      },
    }
  });
  expect(output).toMatchSnapshot();
});

it('serializes objects with keys starting with $ correctly', () => {
  const obj = {
    $id: 1,
    name: 'Alice',
    $type: 'user'
  };
  const output = jsonStringify(obj);
  expect(output).toMatchSnapshot();
});

it('serializes arrays without newlines when length is below the inlineArrayLimit', () => {
  const obj = {
    numbers: [1, 2, 3]
  };
  const options = {
    inlineArrayLimit: 3
  };
  const output = jsonStringify(obj, options);
  expect(output).toMatchSnapshot();
});

it('serializes arrays without newlines when length exceeds the inlineArrayLimit', () => {
  const obj = {
    numbers: [1, 2, 3, 4, 5]
  };
  const options = {
    inlineArrayLimit: 3
  };
  const output = jsonStringify(obj, options);
  expect(output).toMatchSnapshot();
});
it('serializes arrays with newlines when length exceeds the inlineArrayLimit with space set', () => {
  const obj = {
    numbers: [1, 2, 3, 4, 5]
  };
  const options = {
    inlineArrayLimit: 3,
    space: 2
  };
  const output = jsonStringify(obj, options);
  expect(output).toMatchSnapshot();
});

it('serializes arrays without newlines when length equals the inlineArrayLimit', () => {
  const obj = {
    numbers: [1, 2, 3, 4, 5]
  };
  const options = {
    inlineArrayLimit: 5,
    space: 2
  };
  const output = jsonStringify(obj, options);
  expect(output).toMatchSnapshot();
});

it('handles nested arrays with inlineArrayLimit', () => {
  const obj = {
    matrix: [
      [1, 2],
      [3, 4]
    ]
  };
  const options = {
    inlineArrayLimit: 2, // Applies to inner arrays
    space: 2
  };
  const output = jsonStringify(obj, options);
  expect(output).toMatchSnapshot();
});
