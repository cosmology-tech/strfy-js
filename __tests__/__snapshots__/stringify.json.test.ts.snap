// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applies replacer function if provided 1`] = `"[{"firstName": "Alice", "lastName": "Johnson", "age": 20}, {"firstName": "Dana", "lastName": "Johnson", "age": 46}]"`;

exports[`handles complex nested objects with arrays 1`] = `
"{
  "user": {
    "id": 2,
    "details": {
      "name": "Bob",
      "hobbies": [
        "skiing",
        "cycling"
      ]
    }
  }
}"
`;

exports[`handles nested arrays with inlineArrayLimit 1`] = `
"{
  "matrix": [[1, 2], [3, 4]]
}"
`;

exports[`properly escapes strings when necessary 1`] = `"{"complexString": "She said, \\"That's \`incredible\`!\\""}"`;

exports[`serializes arrays with newlines when length exceeds the inlineArrayLimit with space set 1`] = `
"{
  "numbers": [
    1,
    2,
    3,
    4,
    5
  ]
}"
`;

exports[`serializes arrays without newlines when length equals the inlineArrayLimit 1`] = `
"{
  "numbers": [1, 2, 3, 4, 5]
}"
`;

exports[`serializes arrays without newlines when length exceeds the inlineArrayLimit 1`] = `"{"numbers": [1, 2, 3, 4, 5]}"`;

exports[`serializes arrays without newlines when length is below the inlineArrayLimit 1`] = `"{"numbers": [1, 2, 3]}"`;

exports[`serializes objects with keys starting with $ correctly 1`] = `"{"$id": 1, "name": "Alice", "$type": "user"}"`;

exports[`serializes simple objects without quotes on keys where possible 1`] = `"{"id": 1, "name": "Alice", "isActive": true}"`;

exports[`switches to backticks when single quotes are in the string 1`] = `"{"message": "It's a wonderful day!"}"`;

exports[`uses double quotes when backticks and single quotes are present 1`] = `"{"quote": "\`This\` is 'awesome'!"}"`;

exports[`uses single quotes for strings by default 1`] = `"{"greeting": "Hello, world!"}"`;
